### ISSUE TITLE
Trigger custom event 'BlazorLoaded' so we can hook into the event with JavaScript

### ISSUE THREAD

**Issue by Swimburger** [dotnet/aspnetcore#37074]
### Is your feature request related to a problem? Please describe.
I am trying to render Blazor components from JavaScript using the .NET 6 RC1 bits.
This didn't work at first, but [this SO](https://stackoverflow.com/questions/69319216/blazor-enable-dynamic-root-components-error-dynamic-root-components-have-not) answer solved my problem.
If you run `Blazor.rootComponents.add` before Blazor has been loaded, you will get the following error: 
```
Error: Dynamic root components have not been enabled in this application.
```

The solution was to disable autostart using the `autostart="false"` attribute, and then manually start Blazor and when the promise resolves, call the code to add a root component:
```html
    <script src="_framework/blazor.webassembly.js" autostart="false"></script>
    <script>
        Blazor
            .start()
            .then(() => {
                const containerElement = document.querySelector('#app')
                return Blazor.rootComponents.add(containerElement, 'app', {});
            })
            .catch((error) => console.error(error));
    </script>
```

This works fine and is a good solution, but I think it would be helpful [to trigger a custom JavaScript event](https://developer.mozilla.org/en-US/docs/Web/Events/Creating_and_triggering_events) that we can tap into.

### Describe the solution you'd like
I think the following API would be helpful:
```html
    <script src="_framework/blazor.webassembly.js"></script>
    <script>
        window.addEventListener('BlazorLoaded', async function(){
            const containerElement = document.querySelector('#app')
            await Blazor.rootComponents.add(containerElement, 'app', {});
        });
    </script>
```
Or instead of using the browser's native eventing on window or document, add something to the `Blazor` object like this
```html
    <script src="_framework/blazor.webassembly.js"></script>
    <script>
        Blazor.on('loaded', async function(){
            const containerElement = document.querySelector('#app')
            await Blazor.rootComponents.add(containerElement, 'app', {});
        });
    </script>
```

This seems low priority to me, since we do have a solution already, but I do think it would be helpful.

**Comment by user72356** [comment]
Isn't it what the new JavaScript Initializers are for? (https://docs.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/?view=aspnetcore-6.0#javascript-initializers)

**Comment by SQL-MisterMagoo** [comment]
@user72356 Thanks for the pointer - it looks like no to that - although it gave me enough information to get closer.

The new JavaScript Initializer `afterStarted` is called (potentially) too soon for Dynamic components, but it turns out you can pass an initializer to be called for a dynamic component - but it is not well documented or intuitive imho.

To make this work, I changed the app startup like this (adding `javaScriptInitializer: "loadApp"`) in program.cs:

`builder.RootComponents.RegisterForJavaScript<App>(identifier: "app", javaScriptInitializer: "loadApp");`

Then, and this is where it was non-intuitive, I added an export to my module (like `afterStarted`) called `loadApp` - but it wasn't called.

It turned out that `loadApp` is only found if I add it to the `window` object, so I added this to index.html
```html
<script src="_framework/blazor.webassembly.js"></script>
<script>
	window.loadApp = function (component,params) {
		let containerElement = document.querySelector('#app')
		window.Blazor.rootComponents.add(containerElement, component, params)
	}
</script>
```
This feels like I am missing something in how to export my custom initializer OR it really does need to be added to `window` directly, which seems odd...

**Comment by Xyncgas** [comment]
You can put a function called BlazorLoaded in your javascript, then in your blazor when it is actually loaded, call that javascript function from blazor.
If you are looking for the is loaded signal for the runtime of blazor-wasm which is the .net runtime loaded in webassembly, there's api for when a wasm module has loaded.
if you are looking for asp.net core itself being loaded in the browser on top of the .net runtime, currently it is not possible (right?)

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. This is already possible using JavaScript initializers feature that we've announced recently: https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-6-rc-1/#javascript-initializers

**Comment by javiercn** [comment]
@SQL-MisterMagoo can you help me better understand what you mean by this?

> The new JavaScript Initializer `afterStarted` is called (potentially)

In which case are you seeing that this doesn't work.

**Comment by Swimburger** [comment]
Trying to render RootComponents from JS inside the `afterStarted` JavaScript initializer does not work for me.
```js
export function beforeStart(options) {
}
export function afterStarted(blazor) {
    const containerElement = document.querySelector('#app')
    return blazor.rootComponents.add(containerElement, 'app', {});
}
```
It is being called successfully, but the same error occurs: 
```
blazor.webassembly.js:1 Uncaught (in promise) Error: Dynamic root components have not been enabled in this application.
    at E (blazor.webassembly.js:1)
    at Object.add (blazor.webassembly.js:1)
    at afterStarted (RenderBlazorFromJs.lib.module.js:10)
    at blazor.webassembly.js:1
    at Array.map (<anonymous>)
    at It.invokeAfterStartedCallbacks (blazor.webassembly.js:1)
    at Ct (blazor.webassembly.js:1)
```
---

The `window.loadApp` approach does work for me.

---
I think emitting an event through Blazor.on, window.addEventListener, or something similar would be more intuitive than the provided options above.

**Comment by javiercn** [comment]
@Swimburger thanks.

I'm preparing a fix for it. You should be able to do it at that time.

**Comment by SQL-MisterMagoo** [comment]
@javiercn It seems like you got your answer for this, but will your fix include the issue with the javaScriptInitializer in here needing to be declared on `window` instead of in the module js file or does that warrant another issue being opened?

`builder.RootComponents.RegisterForJavaScript<App>(identifier: "app", javaScriptInitializer: "loadApp");`

**Comment by Swimburger** [comment]
@SQL-MisterMagoo my guess is that the parameter `javaScriptInitializer` is unfortunately named the same, but not actually the same as the module JavaScript initializers.

**Comment by javiercn** [comment]
@Swimburger @SQL-MisterMagoo yes, I think the parameter name is a bit unfortunate, it's not clear if we'll be able to change it at this point, but I'm going to create a separate issue for it.

**Comment by javiercn** [comment]
I've filed a separate issue to see if we can deal with the naming appropriately. https://github.com/dotnet/aspnetcore/issues/37181
